### PR TITLE
Add two simple pipeline stages and some tests for the pipeline.

### PR DESCRIPTION
**Pull request by veluca93**
Also fixes multiple bugs in the pipeline implementation and adds various utilities to `image`.

### CODE REVIEW

File: jxl/src/render/stages/convert.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        c == self.channel
+    }
+
+    fn process_row_chunk(
+        &mut self,
+        _position: (usize, usize),
+        xsize: usize,
+        row: &mut [(&[&[u8]], &mut [&mut [f32]])],
+    ) {
+        let (input, output) = &mut row[0];
```
Comment (tirr-c) [inline jxl/src/render/stages/convert.rs:35]: Shouldn't this be `row[self.channel]`? Is it handled by the pipeline runner? (also in `SaveStage`)

File: jxl/src/render/simple_pipeline.rs [hunk 2]
```diff
@@ -85,22 +112,37 @@ impl RenderPipelineBuilder for SimpleRenderPipelineBuilder {
...
+            "final channel info after all stages {:?}",
+            channel_info.last().unwrap()
+        );
         let num_channels = channel_info[0].len();
-        for stage in (0..self.pipeline.stages.len()).rev() {
-            let mut stage_iter = channel_info.iter_mut().skip(stage);
+        for (s, stage) in self.pipeline.stages.iter().enumerate().rev() {
+            let mut stage_iter = channel_info.iter_mut().skip(s);
             let current_info = stage_iter.next().unwrap();
             let next_info = stage_iter.next().unwrap();
```
[the remark this answers is not in the document]
Comment (veluca93) [reply jxl/src/render/simple_pipeline.rs]: nice!

File: jxl/src/render/stages/convert.rs (hunk 1 above)
Comment (veluca93) [reply jxl/src/render/stages/convert.rs:35]: No, this is indeed handled by the pipeline runner.

Comment (tirr-c) [reply jxl/src/render/stages/convert.rs:35]: Ah, I just found the iterator filtering in `do_render` 😅

**Review by sboukortt — APPROVED** [review]

**Pull request merged**